File: src/main/java/org/simpleframework/xml/stream/InputNodeMap.java
/*
 * InputNodeMap.java July 2006
 *
 * Copyright (C) 2006, Niall Gallagher <niallg@users.sf.net>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General 
 * Public License along with this library; if not, write to the 
 * Free Software Foundation, Inc., 59 Temple Place, Suite 330, 
 * Boston, MA  02111-1307  USA
 */

package org.simpleframework.xml.stream;

import javax.xml.stream.events.StartElement;
import javax.xml.stream.events.Attribute;
import java.util.HashMap;
import java.util.Iterator;

/**
 * The <code>InputNodeMap</code> object represents a map to contain
 * attributes used by an input node. This can be used as an empty
 * node map, it can be used to extract its values from a start
 * element. This creates <code>InputAttribute</code> objects for 
 * each node added to the map, these can then be used by an element
 * input node to represent attributes as input nodes.
 *
 * @author Niall Gallagher
 */ 
class InputNodeMap extends HashMap<String, InputNode> implements NodeMap {

   /**
    * This is the source node that this node map belongs to.
    */          
   private InputNode source;        
   
   /**
    * Constructor for the <code>InputNodeMap</code> object. This
    * is used to create an empty input node map, which will create
    * <code>InputAttribute</code> object for each inserted node.
    *
    * @param source this is the node this node map belongs to
    */ 
   protected InputNodeMap(InputNode source) {
      this.source = source;            
   }        

   /**
    * Constructor for the <code>InputNodeMap</code> object. This
    * is used to create an input node map, which will be populated
    * with the attributes from the <code>StartElement</code> that
    * is specified.
    *
    * @param source this is the node this node map belongs to
    * @param element the element to populate the node map with
    */ 
   public InputNodeMap(InputNode source, StartElement element) {
      this.source = source;           
      this.put(element);   
   }

   /**
    * This is used to get the name of the element that owns the
    * nodes for the specified map. This can be used to determine
    * which element the node map belongs to.
    * 
    * @return this returns the name of the owning element
    */         
   public String getName() {
      return source.getName();           
   }   

   /**
    * This is used to insert all attributes belonging to the start
    * element to the map. All attributes acquired from the element
    * are converted into <code>InputAttribute</code> objects so 
    * that they can be used as input nodes by an input node.
    *
    * @param element the element to acquire attributes from
    */ 
   private void put(StartElement element) {
      Iterator list = element.getAttributes();
      
      while(list.hasNext()) {
         Object event = list.next();
         
         if(event instanceof Attribute) {
            put((Attribute)event);                 
         }              
      }           
   }

   /**
    * This is used to insert an <code>Attribute</code> node to 
    * the map. The inserted attribute is converted into an input
    * node by wrapping it in an <code>InputAttribute</code> object.
    * Once the node is inserted it can be acquired by its name.
    *
    * @param event this is the attribute to add to this node map
    */     
   private void put(Attribute event) {
      put(new InputAttribute(source, event));           
   }

   /**
    * This is used to add a new <code>InputAttribute</code> node to
    * the map. The created node can be used by an input node to
    * to represent the attribute as another input node. Once the 
    * node is created it can be acquired using the specified name.
    *
    * @param name this is the name of the node to be created
    * @param value this is the value to be given to the node
    */    
   public void put(String name, String value) {
      put(new InputAttribute(source, name, value));
   }


   /**
    * This is used to insert an <code>InputAttribute</code> node 
    * to the map. The inserted node can be used by an input node to
    * to represent the attribute as another input node. Once the 
    * node is inserted it can be acquired using the attribute name.
    *
    * @param input this is the attribute to add to the node map
    */    
   private void put(InputAttribute input) {
      put(input.getName(), input);
   }
   
   /**
    * This is used to remove the <code>Node</code> mapped to the
    * given name.  This returns a name value pair that represents
    * an attribute. If no node is mapped to the specified name 
    * then this method will return a null value.
    *
    * @param name this is the name of the node to remove
    * 
    * @return this will return the node mapped to the given name
    */    
   public InputNode remove(String name) {
      return super.remove(name);
   }
   
   /**
    * This is used to acquire the <code>Node</code> mapped to the
    * given name. This returns a name value pair that represents
    * an attribute. If no node is mapped to the specified name 
    * then this method will return a null value.
    *
    * @param name this is the name of the node to retrieve
    * 
    * @return this will return the node mapped to the given name
    */       
   public InputNode get(String name) {
      return super.get(name);
   }

   /**
    * This returns an iterator for the names of all the nodes in
    * this <code>NodeMap</code>. This allows the names to be 
    * iterated within a for each loop in order to extract nodes.
    *
    * @return this returns the names of the nodes in the map
    */    
   public Iterator<String> iterator() {
      return keySet().iterator();
   }
}

File: port/src/main/Xml/Stream/InputNodeMap.cs
#region License
//
// InputNodeMap.cs July 2006
//
// Copyright (C) 2006, Niall Gallagher <niallg@users.sf.net>
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied. See the License for the specific language governing
// permissions and limitations under the License.
//
#endregion
#region Using directives
using System.Collections.Generic;
using System;
#endregion
namespace SimpleFramework.Xml.Stream {
   /// <summary>
   /// The <c>InputNodeMap</c> object represents a map to contain
   /// attributes used by an input node. This can be used as an empty
   /// node map, it can be used to extract its values from a start
   /// element. This creates <c>InputAttribute</c> objects for
   /// each node added to the map, these can then be used by an element
   /// input node to represent attributes as input nodes.
   /// </summary>
   class InputNodeMap : LinkedHashMap<String, InputNode> : NodeMap<InputNode> {
      /// <summary>
      /// This is the source node that this node map belongs to.
      /// </summary>
      private readonly InputNode source;
      /// <summary>
      /// Constructor for the <c>InputNodeMap</c> object. This
      /// is used to create an empty input node map, which will create
      /// <c>InputAttribute</c> object for each inserted node.
      /// </summary>
      /// <param name="source">
      /// this is the node this node map belongs to
      /// </param>
      protected InputNodeMap(InputNode source) {
         this.source = source;
      }
      /// <summary>
      /// Constructor for the <c>InputNodeMap</c> object. This
      /// is used to create an input node map, which will be populated
      /// with the attributes from the <c>StartElement</c> that
      /// is specified.
      /// </summary>
      /// <param name="source">
      /// this is the node this node map belongs to
      /// </param>
      /// <param name="element">
      /// the element to populate the node map with
      /// </param>
      public InputNodeMap(InputNode source, EventNode element) {
         this.source = source;
         this.Build(element);
      }
      /// <summary>
      /// This is used to insert all attributes belonging to the start
      /// element to the map. All attributes acquired from the element
      /// are converted into <c>InputAttribute</c> objects so
      /// that they can be used as input nodes by an input node.
      /// </summary>
      /// <param name="element">
      /// the element to acquire attributes from
      /// </param>
      public void Build(EventNode element) {
         for(Attribute entry : element) {
            InputAttribute value = new InputAttribute(source, entry);
            if(!entry.isReserved()) {
               Put(value.Name, value);
            }
         }
      }
      /// <summary>
      /// This is used to acquire the actual node this map represents.
      /// The source node provides further details on the context of
      /// the node, such as the parent name, the namespace, and even
      /// the value in the node. Care should be taken when using this.
      /// </summary>
      /// <returns>
      /// this returns the node that this map represents
      /// </returns>
      public InputNode Node {
         get {
             return source;
         }
      }
      //public InputNode GetNode() {
      //    return source;
      //}
      /// This is used to get the name of the element that owns the
      /// nodes for the specified map. This can be used to determine
      /// which element the node map belongs to.
      /// </summary>
      /// <returns>
      /// this returns the name of the owning element
      /// </returns>
      public String Name {
         get {
            return source.Name;
         }
      }
      //public String GetName() {
      //   return source.Name;
      //}
      /// This is used to add a new <c>InputAttribute</c> node to
      /// the map. The created node can be used by an input node to
      /// to represent the attribute as another input node. Once the
      /// node is created it can be acquired using the specified name.
      /// </summary>
      /// <param name="name">
      /// this is the name of the node to be created
      /// </param>
      /// <param name="value">
      /// this is the value to be given to the node
      /// </param>
      /// <returns>
      /// this returns the node that has just been added
      /// </returns>
      public InputNode Put(String name, String value) {
         InputNode node = new InputAttribute(source, name, value);
         if(name != null) {
            Put(name, node);
         }
         return node;
      }
      /// <summary>
      /// This is used to remove the <c>Node</c> mapped to the
      /// given name.  This returns a name value pair that represents
      /// an attribute. If no node is mapped to the specified name
      /// then this method will return a null value.
      /// </summary>
      /// <param name="name">
      /// this is the name of the node to remove
      /// </param>
      /// <returns>
      /// this will return the node mapped to the given name
      /// </returns>
      public InputNode Remove(String name) {
         return super.Remove(name);
      }
      /// <summary>
      /// This is used to acquire the <c>Node</c> mapped to the
      /// given name. This returns a name value pair that represents
      /// an attribute. If no node is mapped to the specified name
      /// then this method will return a null value.
      /// </summary>
      /// <param name="name">
      /// this is the name of the node to retrieve
      /// </param>
      /// <returns>
      /// this will return the node mapped to the given name
      /// </returns>
      public InputNode Get(String name) {
         return super.Get(name);
      }
      /// <summary>
      /// This returns an iterator for the names of all the nodes in
      /// this <c>NodeMap</c>. This allows the names to be
      /// iterated within a for each loop in order to extract nodes.
      /// </summary>
      /// <returns>
      /// this returns the names of the nodes in the map
      /// </returns>
      public Iterator<String> Iterator() {
         return keySet().Iterator();
      }
   }
}
